modules/client/src/main/kotlin/jp/co/soramitsu/iroha2/SingletonHolder.kt

```diff
@@ -0,0 +1,30 @@
+package jp.co.soramitsu.iroha2
+
+open class SingletonHolder<out T : Any, in A>(creator: (A) -> T) {
+    private var creator: ((A) -> T)? = creator
+
+    @Volatile
+    private var instance: T? = null
+
+    open fun destroy() {
+        instance = null
+    }
+
+    fun getInstance(arg: A): T {
+        val checkInstance = instance
```

Comment:
for sake of simplicity I'd suggest to mark `destroy()` and `getInstance()` (or the class altogether) `synchronized` and remove this block of code

Reply:
in several sources, this singleton implementation is called the most efficient in Kotlin

Reply:
I would not change it